src/ray/common/id.cc

```diff
@@ -172,49 +172,65 @@ uint8_t ObjectID::GetTransportType() const {
 
 // This code is from https://sites.google.com/site/murmurhash/
 // and is public domain.
-uint64_t MurmurHash64A(const void *key, int len, unsigned int seed) {
-  const uint64_t m = 0xc6a4a7935bd1e995;
-  const int r = 47;
-
-  uint64_t h = seed ^ (len * m);
-
-  const uint64_t *data = reinterpret_cast<const uint64_t *>(key);
-  const uint64_t *end = data + (len / 8);
-
-  while (data != end) {
-    uint64_t k = *data++;
-
-    k *= m;
-    k ^= k >> r;
-    k *= m;
+uint64_t MurmurHash64B(const void *key, int len, uint64_t seed) {
+  const uint32_t m = 0x5bd1e995;
+  const int r = 24;
+
+  uint32_t h1 = uint32_t(seed) ^ len;
+  uint32_t h2 = uint32_t(seed >> 32);
+
+  const uint32_t *data = (const uint32_t *)key;
+
+  while (len >= 8) {
+    uint32_t k1 = *data++;
+    k1 *= m;
+    k1 ^= k1 >> r;
+    k1 *= m;
+    h1 *= m;
+    h1 ^= k1;
+    len -= 4;
+
+    uint32_t k2 = *data++;
+    k2 *= m;
+    k2 ^= k2 >> r;
+    k2 *= m;
+    h2 *= m;
+    h2 ^= k2;
+    len -= 4;
+  }
 
-    h ^= k;
-    h *= m;
+  if (len >= 4) {
+    uint32_t k1 = *data++;
+    k1 *= m;
+    k1 ^= k1 >> r;
+    k1 *= m;
+    h1 *= m;
+    h1 ^= k1;
+    len -= 4;
   }
 
-  const unsigned char *data2 = reinterpret_cast<const unsigned char *>(data);
-
-  switch (len & 7) {
-  case 7:
-    h ^= uint64_t(data2[6]) << 48;
-  case 6:
-    h ^= uint64_t(data2[5]) << 40;
-  case 5:
-    h ^= uint64_t(data2[4]) << 32;
-  case 4:
-    h ^= uint64_t(data2[3]) << 24;
+  switch (len) {
   case 3:
-    h ^= uint64_t(data2[2]) << 16;
+    h2 ^= ((unsigned char *)data)[2] << 16;
```

Comment:
Can we use a C++ cast here instead? I realize it's longer, but it's less error-prone. I think `static_cast` should suffice here. (Also, `const` shouldn't be cast away.)